File: desktop/src/Beam.Desktop/Services/ImageFingerprint.cs
using System.IO;
using System.Security.Cryptography;
using System.Windows.Media;
using System.Windows.Media.Imaging;

namespace Beam.Desktop.Services;

/// <summary>
/// Impressão digital de imagem: SHA-256 dos pixels de uma miniatura 64×64 (BGR24).
/// Como o PNG é sem perdas e o redimensionamento do WPF é determinístico, a MESMA
/// imagem produz o MESMO hash mesmo após decodificar/recodificar — ao contrário dos
/// bytes do arquivo, que mudam a cada recodificação. É isso que permite detectar
/// "esta imagem já foi sincronizada" e cortar o loop infinito de eco, sem colidir
/// entre imagens diferentes.
/// </summary>
public static class ImageFingerprint
{
    private const int Size = 64;

    public static ulong Compute(byte[] pngBytes)
    {
        using var stream = new MemoryStream(pngBytes);
        var decoder = BitmapDecoder.Create(stream, BitmapCreateOptions.None, BitmapCacheOption.OnLoad);
        var frame = decoder.Frames[0];

        var scaled = new TransformedBitmap(
            frame,
            new ScaleTransform((double)Size / frame.PixelWidth, (double)Size / frame.PixelHeight));
        // BGR24 descarta o canal alfa: o round-trip pelo clipboard (DIB) pode
        // alterá-lo, mas mantém RGB intacto.
        var canonical = new FormatConvertedBitmap(scaled, PixelFormats.Bgr24, null, 0);

        var width = canonical.PixelWidth;
        var height = canonical.PixelHeight;
        var stride = width * 3;
        var pixels = new byte[stride * height];
        canonical.CopyPixels(pixels, stride, 0);

        var digest = SHA256.HashData(pixels);
        return BitConverter.ToUInt64(digest, 0);
    }

    public static bool IsSameImage(ulong a, ulong b) => a == b;
}

File: mobile/app/src/main/java/com/esousa/beam/clipboard/ImageFingerprint.kt
package com.esousa.beam.clipboard

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import java.nio.ByteBuffer
import java.security.MessageDigest

/**
 * Impressão digital de imagem: SHA-256 dos pixels RGB de uma miniatura 64×64.
 * Como o PNG é sem perdas e o redimensionamento é determinístico, a MESMA imagem
 * produz o MESMO hash mesmo após decodificar/recodificar — é o que permite detectar
 * "esta imagem já foi sincronizada" e cortar o loop infinito de eco, sem colidir
 * entre imagens diferentes.
 */
object ImageFingerprint {
    private const val SIZE = 64

    fun compute(pngBytes: ByteArray): Long? {
        val bitmap = BitmapFactory.decodeByteArray(pngBytes, 0, pngBytes.size) ?: return null
        return try {
            compute(bitmap)
        } finally {
            if (!bitmap.isRecycled) bitmap.recycle()
        }
    }

    fun compute(bitmap: Bitmap): Long {
        val small = Bitmap.createScaledBitmap(bitmap, SIZE, SIZE, true)
        try {
            val pixels = IntArray(SIZE * SIZE)
            small.getPixels(pixels, 0, SIZE, 0, 0, SIZE, SIZE)

            // Somente RGB — o alfa pode ser alterado por conversões intermediárias.
            val rgb = ByteBuffer.allocate(pixels.size * 3)
            pixels.forEach { pixel ->
                rgb.put(((pixel shr 16) and 0xFF).toByte())
                rgb.put(((pixel shr 8) and 0xFF).toByte())
                rgb.put((pixel and 0xFF).toByte())
            }

            val digest = MessageDigest.getInstance("SHA-256").digest(rgb.array())
            return ByteBuffer.wrap(digest, 0, 8).long
        } finally {
            if (small !== bitmap && !small.isRecycled) small.recycle()
        }
    }

    fun isSameImage(a: Long, b: Long): Boolean = a == b
}
